src/ucs/sys/event_set.c

```diff
@@ -0,0 +1,166 @@
+}
+
+ucs_status_t ucs_event_set_create(ucs_sys_event_set_t **event_set_p)
+{
+    ucs_sys_event_set_t *event_set;
+    ucs_status_t status;
+
+    event_set = ucs_malloc(sizeof(ucs_sys_event_set_t), "ucs_sys_event_set");
+    if (event_set == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out_create;
+    }
+
+    /* Create epoll set the thread will wait on */
+    event_set->epfd = epoll_create(1);
+    if (event_set->epfd < 0) {
+        ucs_error("epoll_create() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_free;
+    }
+
+    status = UCS_OK;
+    *event_set_p = event_set;
+    goto out_create;
+
+err_free:
+    ucs_free(event_set);
+out_create:
+    return status;
+}
+
+ucs_status_t ucs_event_set_add(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+
+    memset(&event, 0, sizeof(event));
+    event.events   = os_events;
+    event.data.fd  = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_ADD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_mod(ucs_sys_event_set_t *event_set, int event_fd,
+                               ucs_event_set_type_t events, void *udata)
+{
+    struct epoll_event event;
+    int ret;
+    int os_events;
+
+    os_events = ucs_event_set_events_map(events);
+
+    memset(&event, 0, sizeof(event));
+    event.events   = os_events;
+    event.data.fd  = event_fd;
+    event.data.ptr = udata;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_MOD, event_fd, &event);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, MOD, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+{
+    int ret;
+
+    ret = epoll_ctl(event_set->epfd, EPOLL_CTL_DEL, event_fd, NULL);
+    if (ret < 0) {
+        ucs_error("epoll_ctl(epfd=%d, DEL, fd=%d) failed: %m", event_set->epfd,
+                  event_fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set, int timeout_ms,
+                                event_set_handler_t event_set_handler)
+{
+    int nready;
+    struct epoll_event events[UCS_EVENT_EPOLL_MAX_EVENTS];
+
+    nready = epoll_wait(event_set->epfd, events, UCS_EVENT_EPOLL_MAX_EVENTS,
+                        timeout_ms);
+    if ((nready < 0) && (errno != EINTR)) {
+        ucs_error("epoll_wait() failed: %m");
+        return UCS_ERR_IO_ERROR;
+    } else if (nready==0) {
+        return UCS_ERR_TIMED_OUT;
```

Comment:
why do we need to return error? just not call any callback..

Reply:
Removed